app/root.tsx

```diff
@@ -137,11 +138,19 @@ function Document({
   );
 }
 
+const matchHandleSchema = z
+  .object({
+    forceDark: z.boolean(),
+  })
+  .optional();
+
```

Comment:
Sorry, I should have been more clear in the issue. I'd prefer not to ship `zod` to the client, since it's [kind of large](https://bundlephobia.com/package/zod@3.22.4).

All good to use it on the server as we already are. This is kind of the challenge with fullstack frameworks that try to blend the client and the server, it's easy to leak stuff like this. Would probably be good to put all zod validation stuff inside of a directory called `validate.server/` or something

Here's a simple, type-safe validation that doesn't require sending `zod` to the client. I get it's a bit verbose, so you're welcome to extract it to a function or figure out a better approach

```js
let forceDark = matches.some(({ handle }) => {
  if (handle && typeof handle === "object" && "forceDark" in handle) {
    return handle.forceDark;
  }
  return undefined;
});
```

Reply:
If you're using Vite, then you can use the `serverOnly$` export from `vite-env-only` to ensure that zod is not included in the client bundle.

```ts
// routes/test.$id.tsx

import { json, type LoaderFunctionArgs } from '@remix-run/node'
import { useLoaderData } from '@remix-run/react'
import { serverOnly$ } from 'vite-env-only'
import { z } from 'zod'

// this ensures that zod is only included in the server bundle
const schema = serverOnly$(() => z.object({ id: z.coerce.number() }))

export async function loader({ params }: LoaderFunctionArgs) {
  const { id } = schema!().parse(params)
  return json({ id })
}

export default function Component() {
  const loaderData = useLoaderData<typeof loader>()

  return (
    <div>
      <pre>{JSON.stringify(loaderData, null, 2)}</pre>
    </div>
  )
}
```